pkg/virtual/apiexport/controllers/apireconciler/apiexport_apireconciler_controller.go

```diff
@@ -0,0 +1,231 @@
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	clusterName, name := clusters.SplitClusterAwareKey(key)
+	exports, err := c.apiExportIndexer.ByIndex(byWorkspace, clusterName.String())
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	if len(exports) == 0 {
+		klog.V(2).Infof("No kubernetes APIExport found for APIResourceSchema %s|%s", clusterName, name)
+		return
+	}
+
+	for _, obj := range exports {
+		export := obj.(*apisv1alpha1.APIExport)
+		klog.V(2).Infof("Queueing APIExport %s|%s for APIResourceSchema %s", clusterName, export.Name, name)
+		c.enqueueAPIExport(obj)
+	}
+}
+
+func (c *APIReconciler) enqueueAPIExport(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *APIReconciler) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *APIReconciler) Start(ctx context.Context) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s controller", ControllerName)
+	defer klog.Infof("Shutting down %s controller", ControllerName)
+
+	go wait.Until(func() { c.startWorker(ctx) }, time.Second, ctx.Done())
+
+	// stop all watches if the controller is stopped
+	defer func() {
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
+		for _, sets := range c.apiSets {
+			for _, v := range sets {
+				v.TearDown()
+			}
+		}
+	}()
+
+	<-ctx.Done()
+}
+
+func (c *APIReconciler) ShutDown() {
+	c.queue.ShutDown()
+}
+
+func (c *APIReconciler) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%s: failed to sync %q, err: %w", ControllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *APIReconciler) process(ctx context.Context, key string) error {
+	clusterName, apiExportName := clusters.SplitClusterAwareKey(key)
+	apiDomainKey := dynamiccontext.APIDomainKey(clusterName.String() + "/" + apiExportName)
+
+	apiExport, err := c.apiExportLister.Get(key)
+	if err != nil && !apierrors.IsNotFound(err) {
+		klog.Errorf("failed to get APIExport %s|%s from lister: %v", clusterName, apiExportName, err)
+		return nil // nothing we can do here
+	}
+
+	return c.reconcile(ctx, apiExport, apiDomainKey)
+}
+
+func (c *APIReconciler) GetAPIDefinitionSet(_ context.Context, key dynamiccontext.APIDomainKey) (apidefinition.APIDefinitionSet, bool, error) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+
+	apiSet, ok := c.apiSets[key]
+	return apiSet, ok, nil
```

Comment:
If you cached errors by key, this is where you could expose them

Reply:
Why expose ?